File: ShortestJobFirst.cpp
// File Name: FirstComeFirstServe.cpp
// Date: December 15, 2022
// Time: 01:10:17
// Author : Sajid (git : sajidshahriar72543)

/*
    Idea is to sort the processes according to their burst time then apply FCFS
*/

#include <bits/stdc++.h>
using namespace std;
void withoutArrival(int n);
void withArrival(int n);
int main()
{
    cout << "Enter no of processes: ";
    int n;
    cin >> n;
    withoutArrival(n);
    // return 0;
}
void withoutArrival(int n)
{
    int bt[n], wt[n], tat[n];
    cout << "Enter burst time of processes: ";
    for (int i = 0; i < n; i++)
    {
        cin >> bt[i]; // input burst time
    }

    for (int i = 0; i < n; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            if (bt[i] > bt[j])
            {
                swap(bt[i], bt[j]);
            }
        }
    }

    wt[0] = 0; // waiting time of first process is 0
    for (int i = 1; i < n; i++)
    {
        wt[i] = bt[i - 1] + wt[i - 1]; // waiting time of other processes
    }
    for (int i = 0; i < n; i++)
    {
        tat[i] = bt[i] + wt[i]; // turn around time of processes
    }
    cout << "Process\tBT\tWT\tTAT" << endl;
    for (int i = 0; i < n; i++)
    {
        cout << i + 1 << "\t" << bt[i] << "\t" << wt[i] << "\t" << tat[i] << endl;
    }
    float sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += wt[i]; // sum of waiting time
    }
    cout << "Average waiting time: " << float(sum / n) << endl;
    sum = 0;
    for (int i = 0; i < n; i++)
    {
        sum += tat[i]; // sum of turn around time
    }
    cout << "Average turn around time: " << float(sum / n);
}

File: ShortestJobFirst.py
# File Name: ShortestJobFirst.py
# Date: 2022-12-15
# Date: December 15, 2022
# Time: 17:09:45
# Author : Sajid (git : sajidshahriar72543)

# Idea is to sort the processes according to their burst time and then execute them in that order
# This is a non-preemptive algorithm, tha is why arrival time is not considered
# This is just FCFS with a sorted list of processes

n = int(input("Enter the number of processes: "))
top = 0
bt = []
ct = []
tat = []
wt = []
avgwt = 0
avgtat = 0
for i in range (n):
    inp = int(input("Enter the burst time of process " + str(i+1) + ": "))
    bt.append(inp)
    print()

for i in range (n):
    if i==0: #first process
        top += bt[i] 
        ct.append(top) #completion time
    elif i>0: #other processes
        top += bt[i] 
        ct.append(top) #completion time

print("Burst Time Calculated\n")

for i in range (n):
    temp = ct[i] #- at[i]
    tat.append(temp) #turn around time
    temp = tat[i] - bt[i]
    wt.append(temp) #waiting time

print("Turn Around Time Calculated\n Waiting Time Calculated\n")

print("PID\tBT\tCT\tTAT\tWT")
for i in range (n):
    print(str(i+1) + "\t" + str(bt[i]) + "\t" + str(ct[i]) + "\t" + str(tat[i]) + "\t" + str(wt[i]))
    avgwt += wt[i]
    avgtat += tat[i]

avgwt = avgwt/n
avgtat = avgtat/n

print("Average Waiting Time: " + str(avgwt))
print("Average Turn Around Time: " + str(avgtat))